runtime/lua/vim/scripts.lua

```diff
@@ -0,0 +1,359 @@
+    function()
+      -- Gprof (gnu profiler)
+      if lines[1] == 'Flat profile:' and lines[2] == '' and lines[3]:find('^Each sample counts as .* seconds%.$') then
+        return 'gprof'
+      end
+    end,
+    -- Erlang terms
+    -- (See also: http://www.gnu.org/software/emacs/manual/html_node/emacs/Choosing-Modes.html#Choosing-Modes)
+    ['%-%*%-.*erlang.*%-%*%-'] = { 'erlang', { ignore_case = true } },
+    -- YAML
+    ['^&YAML'] = 'yaml',
+    -- MikroTik RouterOS script
+    ['^#.*by RouterOS'] = 'routeros',
+    -- Sed scripts
+    -- #ncomment is allowed but most likely a false positive so reuire a space
+    -- before any trailing comment text
+    ['^#n%s'] = 'sed',
+    ['^#n$'] = 'sed',
+  }
+
+  for k, v in pairs(patterns) do
+    if type(v) == 'string' then
+      -- Check the first line only
+      if lines[1]:find(k) then
+        return v
+      end
+    elseif type(v) == 'function' then
+      local ft = v()
+      if ft then
+        return ft
+      end
+    else
+      local opts = type(v) == 'table' and v[2] or {}
+      if opts.start_lnum then
+        assert(not opts.ignore_case, 'ignore_case=true is ignored when start_lnum is also present, needs refactor')
+        for _, line in ipairs(getlines(bufnr, opts.start_lnum, opts.end_lnum)) do
+          if line:find(k) then
+            return v[1]
+          end
+        end
+      else
+        -- Check the first line only
+        local line = opts.ignore_case and lines[1]:lower() or lines[1]
+        if opts.vim_regex and matchregex(line, k) or line:find(k) then
+          return v[1]
+        end
+      end
+    end
+  end
+
+  -- CVS diff
+  for _, line in ipairs(getlines(bufnr, 1, -1)) do
+    if not line:find('^%? ') then
+      if matchregex(line, [[^Index:\s\+\f\+$]]) then
+        return 'diff'
+      elseif
+        -- Locale input files: Formal Definitions of Cultural Conventions
+        -- Filename must be like en_US, fr_FR@euro or en_US.UTF-8
+        findany(path, { '%a%a_%a%a$', '%a%a_%a%a[%.@]', '%a%a_%a%ai18n$', '%a%a_%a%aPOSIX$', '%a%a_%a%atranslit_' })
+      then
+        for _, line_ in ipairs(getlines(bufnr, 1, 100)) do
+          if
+            findany(line_, {
+              '^LC_IDENTIFICATION$',
+              '^LC_CTYPE$',
+              '^LC_COLLATE$',
+              '^LC_MONETARY$',
+              '^LC_NUMERIC$',
+              '^LC_TIME$',
+              '^LC_MESSAGES$',
+              '^LC_PAPER$',
+              '^LC_TELEPHONE$',
+              '^LC_MEASUREMENT$',
+              '^LC_NAME$',
+              '^LC_ADDRESS$',
+            })
+          then
+            return 'fdcc'
+          end
+        end
+      end
+    end
+  end
+end
+
+local cpo_save = vim.o.cpoptions
+vim.cmd('set cpo&vim')
+
+local path = vim.fn.expand('<afile>')
+local bufnr = vim.api.nvim_get_current_buf()
+
+local first_line = getlines(bufnr, 1)
+local ft = first_line:find('^#!') and scripts_shebang(path, bufnr) or scripts_no_shebang(path, bufnr)
```

Comment:
Both `scripts_shebang` and `scripts_no_shebang` also call `getlines(bufnr, 1)`, so you're doing the same work 3 times. Maybe instead pass the first line as an argument into `scripts_shebang` and `scripts_no_shebang`?

Reply:
Good point, I added it for `scripts_shebang` (not for `scripts_no_shebang` because that function also retrieves lines 2 - 5 so I don't think it's needed)